### ISSUE TITLE
mdtraj_refractor: Format for data inputs and outputs

### ISSUE THREAD

**Issue by TensorDuck** [ajkluber/model_builder#5]
I've added a comment to inputs.py to this effect. Not sure if it's the best place to put it, so I'm willing to discuss.

I wanted to formalize our data input and output strategy, and leave comment documentation in the docstrings to that effect. I propose, and I think this is reasonable is:

Input files should be formatted the way they are externally. Think pdb files list residues and atom indices starting from 1 instead of 0. 
Internal methods should assume you are doing the python thing that starts from zero. This is in line with how mdtraj handles their calls to lists and such.

inputs.py should be the main place for these conversions to take place. And the conversions that are automatically be done is stated in the docstrings so users will expect this and not accidentally pre-format their files and shift atom contacts by one for example. Not sure if more examples will occur, but, we should formally state this standard and follow it so that we don't have needless built in checks, and future developers won't have to comb through the code to look for a -1 somewhere. 

Please comment if you agree or disagree.

**Comment by ajkluber** [comment]
I agree that the atom indexing of input files (structure files and params files) should start at 1. I believe this is the case right now. Is there some inconsistency that you have noticed or difficulty that you foresee? 

One potentially confusing point regarding indexing is the atom indexing in params files. The current indexing in `pairwise_params` corresponds to the atom indexing in the mapped topology. So each `pairwise_params` needs to be associated with a mapped topology to make sense. I'm not sure if there is a way to avoid this.

I'm not sure if this is related to you original issue but in regards to formatting inputs: we should decide on a general way to specify the complete set of potentials (bonds, angles, etc) which should also allow for  references to tabulated interactions (e.g. transferable dihedral potential).

**Comment by TensorDuck** [comment]
Currently the case, I wanted to leave a docstring in one of the opening files for future developers to come across easily and use. Could be in the **init** file or in the header of the inputs.py file. I left a comment in the inputs.py file under the load_model method, but that's probably not the best place for it. 

I saw the issue with the params file as well, and I have no solution unfortunately. I think we need to assume that the user has some knowledge of what they are doing. But one way to help them could be to write a generate_inputs package, such as one that has helper functions for writing a pairwise_param file given some typical inputs (residue contacts, contacts specified between atoms in an all-atom pdb, etc.). If we integrate the packages well, at least future users can figure it out easily.

Not sure what you had in mind for your final point. It sounds like you want to have a way to input/output what terms for an angle interaction are and how strong their interaction is?

**Comment by ajkluber** [comment]
I think mentioning indexing in a docstring somewhere is appropriate. Maybe `inputs.py` is the best place for it.

To clarify my other point: there should be a way to save/load an arbitrary model object from files (not just structure-based models). So in addition to `pairwise_params` we have files that specify the backbone interactions (e.g. `bond_params`, `angle_params`, `dihedral_params`). This makes things a little bloated with files, so there may be a more elegant solution. 

What I have in mind is a compact and comprehensive description of a Model object. Ideally this would contain topology information as well as a complete description of the Hamiltonian. (Ideas that are similar in spirit are: 1) the [HDF5 file format](http://mdtraj.org/latest/hdf5_format.html) that mdtraj has proposed for trajectories; 2) the XML file format used by SMOG2 and Hoomd). Maybe we should consider making this a milestone for the future?

A separate but related point that I was trying to raise is that we should have the ability to specify arbitrary tabled interactions in the aforementioned `_params` files. I anticipate that we want the freedom to use more transferable interactions in the future  (e.g. flavored pair or backbone interactions) and implementing them in table files allows for flexible functional forms.

**Comment by ajkluber** [comment]
@TensorDuck do you have code that saves new `pairwise_params` and `model_params` as part of your fitting code? Could you add that functionality to `model_builder` if possible? Probably `models/output` would be the relevant place.

**Comment by TensorDuck** [comment]
Yes, I do have functionality for that. I'll update when I have it implemented.

**Comment by TensorDuck** [comment]
I have implemented a rudimentary class for outputting internal information about a model. 

I've tested, and it looks like the pairwise params file and model params file that goes in is the one that comes out again, so it's at least consistent for now. I haven't tested the full range of possible functions, but we keep an eye on it as it gets used by more people. I'll also add some comments in a latter commit.

Also, I wanted to point out that right now, every `PairPotential` object has a `self.other_params` attribute now, which basically stores the parameters in a list, which then gets called later to output. This works for now, and there aren't too many places that had to be inserted. Maybe this is a problem if that file grows larger as more potentials are added.

I feel like there should be a better more generic way to do stuff like this in the future, but it might take a refactor of the potentials. Was going to think about it more, but I would be open to suggestions. I guess it would be best to leave it as is now, but if we start adding more features to the `PairPotential` objects, we need to consider what to do about that.
